lib/Optimizer/GILPasses/CopyLoweringPass.cpp

```diff
@@ -0,0 +1,131 @@
+#include "GIL/InstVisitor.hpp"
+#include "GIL/Module.hpp"
+#include "GILGen/Context.hpp"
+#include "Instructions/LoadInst.hpp"
+#include "Instructions/ReturnInst.hpp"
+#include "PassManager.hpp"
+
+namespace glu::optimizer {
+
+class CopyLoweringPass : public gil::InstVisitor<CopyLoweringPass> {
+private:
+    gil::Module *module;
+    std::optional<gilgen::Context> ctx = std::nullopt;
+    llvm::BumpPtrAllocator &arena;
+    llvm::SmallVector<gil::InstBase *, 8> toErase;
+    bool inCopyFunction = false; // Track if we're inside a copy function
+
+public:
+    CopyLoweringPass(gil::Module *module, llvm::BumpPtrAllocator &arena)
+        : module(module), arena(arena)
+    {
+    }
+
+    ~CopyLoweringPass()
+    {
+        for (auto *inst : toErase) {
+            inst->eraseFromParent();
+        }
+    }
+
+    void visitLoadInst(gil::LoadInst *loadInst)
+    {
+        // Don't transform loads inside copy functions (would cause infinite
+        // recursion)
+        if (inCopyFunction)
+            return;
```

Comment:
you would need `copy` functions to use a "borrowed" reference so that ownership isn't handled. It might have to be done from a pointer (which means you would only have `load [copy]` and no copy instructions?).